src/Analyzers/Orchestration/OrchestrationAnalyzer.cs

```diff
@@ -0,0 +1,249 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System.Collections.Concurrent;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Operations;
+
+namespace Microsoft.DurableTask.Analyzers.Orchestration;
+
+public abstract class OrchestrationAnalyzer : DiagnosticAnalyzer
+{
+    public override void Initialize(AnalysisContext context)
+    {
+        context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+        context.EnableConcurrentExecution();
+
+        context.RegisterCompilationStartAction(context =>
+        {
+            var knownSymbols = new KnownTypeSymbols(context.Compilation);
+
+            if (knownSymbols.FunctionOrchestrationAttribute == null || knownSymbols.FunctionNameAttribute == null ||
+                knownSymbols.TaskOrchestratorInterface == null || knownSymbols.TaskOrchestratorBaseClass == null ||
+                knownSymbols.DurableTaskRegistry == null)
+            {
+                // symbols not available in this compilation, skip analysis
+                return;
+            }
+
+            IMethodSymbol runAsyncTaskOrchestratorInterface = knownSymbols.TaskOrchestratorInterface.GetMembers("RunAsync").OfType<IMethodSymbol>().First();
+            IMethodSymbol runAsyncTaskOrchestratorBase = knownSymbols.TaskOrchestratorBaseClass.GetMembers("RunAsync").OfType<IMethodSymbol>().First();
+
+            OrchestrationAnalysisResult result = new();
+
+            // look for Durable Functions Orchestrations
+            context.RegisterSyntaxNodeAction(ctx =>
+            {
+                ctx.CancellationToken.ThrowIfCancellationRequested();
+
+                if (ctx.ContainingSymbol is not IMethodSymbol methodSymbol)
+                {
+                    return;
+                }
+
+                if (!methodSymbol.ContainsAttributeInAnyMethodArguments(knownSymbols.FunctionOrchestrationAttribute))
+                {
+                    return;
+                }
+
+                if (!methodSymbol.TryGetSingleValueFromAttribute(knownSymbols.FunctionNameAttribute, out string functionName))
+                {
+                    return;
+                }
+
+                var orchestration = new AnalyzedOrchestration(functionName);
```

Comment:
nit: prefer this style for ctors:

`AnalyzedOrchestration orchestration = new(functionName);`

Reply:
done :)